packages/backend/server/src/plugins/copilot/session.ts

```diff
@@ -170,16 +211,138 @@ export class ChatSessionService {
     const state = await this.cache.get<ChatSessionState>(
       `${CHAT_SESSION_KEY}:${sessionId}`
     );
-    if (state) return state;
+    if (state) {
+      state.messages.forEach(m => {
+        if (typeof m.createdAt === 'string') {
+          m.createdAt = new Date(m.createdAt);
+        }
+      });
+      return state;
+    }
     return await this.getSession(sessionId);
   }
 
+  private calculateTokenSize(
+    messages: ChatMessage[],
+    model: AvailableModel
+  ): number {
+    const encoder = getTokenEncoder(model);
+    return messages.reduce((total, m) => {
+      return total + encoder.encode_ordinary(m.content).length;
+    }, 0);
+  }
+
+  async countSessions(
+    userId: string,
+    workspaceId: string,
+    options?: { docId?: string; action?: boolean }
+  ): Promise<number> {
+    // NOTE: only used for anonymous session in development
+    if (!userId && AFFiNE.featureFlags.copilotAuthorization) {
+      return this.unsavedSessions.size;
+    }
+    return await this.db.aiSession.count({
+      where: {
+        userId,
+        workspaceId,
+        docId: workspaceId === options?.docId ? undefined : options?.docId,
+        action: options?.action,
+      },
+    });
+  }
+
+  async listSessions(
+    userId: string | undefined,
+    workspaceId: string,
+    options?: { docId?: string; action?: boolean }
+  ): Promise<string[]> {
+    // NOTE: only used for anonymous session in development
+    if (!userId && AFFiNE.featureFlags.copilotAuthorization) {
```

Comment:
remember to remove

Reply:
could you move all authentication ignore code to a separated PR? we will easily delete them all by reverting it.

Reply:
lgtm

Reply:
moved to https://github.com/toeverything/AFFiNE/pull/6445